torchdynamo/guards.py

```diff
@@ -50,6 +60,98 @@
 )
 
 
+class NNModuleChangeTrackerUtil:
+    @staticmethod
+    def invalidate_module(module):
+        if module not in module_code_map:
+            # Module is out of scope / deleted (weak ref key ref dict)
+            return
+
+        for code in module_code_map[module]:
+            code.valid = False
+
+        del module_code_map[module]
+
+        if module not in module_to_guard_ids:
+            # Module is out of scope / deleted (weak ref key ref dict)
+            return
+
+        for object_id in module_to_guard_ids[module]:
+            module_associated_guarded_ids.remove(object_id)
+
+        del module_to_guard_ids[module]
+
+    @staticmethod
+    def setup(module, guarded_code):
+        module_id = id(module)
+        modulecls = module.__class__
+        if module_id not in module_to_guard_ids:
+            module_to_guard_ids[module] = set()
+
+        for buffer in module.buffers():
+            buffer_id = id(buffer)
+            module_associated_guarded_ids.add(buffer_id)
+            module_to_guard_ids[module].add(buffer_id)
+
+        for mod in module.modules():
+            mod_id = id(mod)
+            module_associated_guarded_ids.add(mod_id)
+            module_to_guard_ids[module].add(mod_id)
+
+        for parameter in module.parameters():
+            parameter_id = id(parameter)
+            module_associated_guarded_ids.add(parameter_id)
+            module_to_guard_ids[module].add(parameter_id)
+
+        if getattr(modulecls, "__dynamo_module_patch", True):
```

Comment:
If you have a class hierarchy, `A->B->C->D->nn.Module`:
I think it is possible we patch each class in the hierarchy and end up getting multiple calls to `invalidate_module`.

Perhaps we should just patch this single torch.nn.Module base class.

Reply:
Yea, I saw that, but I wasn't sure if I'd miss any edge cases with this. I'll simplify it down to just nn.Module and see how it goes.